File: 1_course/week3/6_max_num_prizes/python/main.py
# python3

##
#
# Python3 implementation to find the maximum number of prizes
#
# (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
#
##

from typing import List

class Solution:
    def maxNumPrizes( self, N: int ) -> List[int] :
        ans = []
        i = 1
        while i+1 <= N-i:
            ans.append( i )
            N -= i
            i += 1
        if 0 < N:
            ans.append( N )
        return ans

if __name__ == '__main__':
    solution = Solution()
    N = int( input() )
    ans = solution.maxNumPrizes( N )
    print( len( ans ))
    for x in ans:
        print( x, end=" " )

File: 1_course/week3/6_max_num_prizes/java/src/Main.java
/**
 *
 * Java implementation to find the maximum number of prizes
 *
 * (c) Copyright 2019 Clayton J. Wong ( http://www.claytonjwong.com )
 *
 **/

import java.util.Scanner;
import java.util.ArrayList;

public class Main {

    public static ArrayList maxNumPrizes( int N ){
        ArrayList ans = new ArrayList();
        for( int i=1; i+1 <= N-i; N-=i, ++i )
            ans.add( i );
        if( 0 < N )
            ans.add( N );
        return ans;
    }

    public static void main( String[] args ){
        Scanner input = new Scanner( System.in );
        int N = input.nextInt();
        ArrayList ans = maxNumPrizes( N );
        System.out.println( ans.size() );
        ans.forEach(( i ) -> System.out.print( i + " " ));
    }
}
